### ISSUE TITLE
Support busy response for RPCs in iotile-emulate

### ISSUE THREAD

**Issue by timburke** [iotile/coretools#619]
We support async RPCs, which means the emulator should block rpcs to tiles with a busy response if the tile is currently executing an async rpc.  Currently we default to allowing rpcs to run on a tile while an async rpc is pending.

**Comment by DimaWittmann** [comment]
Hi Tim! 

I thought to investigate this ticket. Could you, please, say if you don't mind? I saw you made a lot of changes. 

In case you don't mind I'd like to ask you also a few questions. As I understand proxies implement rpc synchronously, so it's not possible to call two async rpcs through a proxy. Am I right? I just trying to understand how to issue another rpc while one is pending.
Another thing I haven't yet understand how is busy status should be reported. I'll dive more into code but in case you have any thoughts I would be happy if you share.

Thank you,
Dmytrii

**Comment by timburke** [comment]
Hi Dmytrii,

The busy RPC response is returned by a tile if another RPC is received by that tile while the tile has an asynchronous RPC pending.  You are correct that from the caller's point of view all RPCs are synchronous, so a single caller can never receive a busy response since their call to `rpc` will not return until the RPC has finished, regardless of whether the underlying implementation chose to perform the RPC work in the rpc method implementation itself (the method decorated with `@tile_rpc`) or whether it chose to perform the work asynchronously and return the response via a callback.

Key facts to understand:

- There is a single global rpc dispatch queue for the entire device.  This means that exactly one RPC handler can be executing in the device at one.  In a physical device this is achieved by using mutual exclusion on shared physical bus lines that are used to communicate between the tiles.  In the emulator, it is achieved with a single rpc_dispatcher coroutine task that has a single global queue of rpcs and it dispatches them one-by-one to tiles.

- when an `asynchronous` rpc is in progress (after the initial decorated method has raised `AsynchronousRPCResponse` and before it finishes the RPC by calling `finish_async_rpc`, the RPC is still in progress on that tile but the shared bus is free so RPCs can be executed on **other tiles** by **other callers**.  This is the primary benefit of asynchronous RPCs.  They allow long-running RPCs to do their work without locking the global RPC bus for the entire device and preventing other RPCs from running.

   Basically, the difference between a synchronous and asynchronous RPC is whether the bus is locked during the entire RPC (synchronous) or the bus is released during RPC execution and the tile implementing the RPC calls back the caller tile when it is done (so there are two bus sessions with an arbitrary amount of time in between).  During the intermediate time when the bus is free, other RPCs by other callers may run.

- due to resource constraints in physical devices and simplicity of reasoning, a single tile is only allowed to have at most 1 rpc in progress at a given time, so once it responds with `AsynchronousRPCResponse` to a single RPC, it must fail all future RPCs with `BusyResponse` until the asynchronous RPC is finished.

The busy response to an RPC can come in the following situation:
- caller A sends RPC X to tile 1.  Tile 1 chooses to handle the RPC asynchronously.  This frees the RPC dispatcher to handle other RPCs while RPC X is running on tile 1.
- caller B sends RPC Y to tile 1.  RPC Y **must fail immediately with BusyResponse** regardless of whether RPC Y would normally be implemented synchronously or asynchronously.  The busy response is the same whether Y==X or Y!=X, so any RPC on tile 1 will be failed as busy while the asynchronous RPC is pending.

In contrast the following situation would have no busy response:
- caller A sends RPC X to tile 1.  Tile 1 chooses to handle the RPC asynchronously. 
- caller B sends RPC Y to tile 2.  Since this is sent to a different tile, it proceeds without being disturbed by the pending RPC on tile 1.  RPC Y can be handled either synchronously or asynchronously with no limitation.

@DimaWittmann Please let me know if any of this is unclear.

**Comment by timburke** [comment]
@DimaWittmann Note that you cannot get a busy response from a proxy unless the device is sending automated RPCs inside of itself at the same time since your single proxy will never be able to have 2 rpcs in flight at once.

**Comment by timburke** [comment]
@DimaWittmann Note as well that for the purposes of `iotile-emulate` an RPC that is implemented as a coroutine with an `await def rpc_handler()` definition and an `@async_tile_rpc` decorator is considered to be **synchronous** from the point of view of this discussion and the global RPC queue blocks until it finishes.  The only RPCs that are considered asynchronous are those that raise `AsynchronousRPCResponse`.

The point of the `@async_tile_rpc` is to allow for the simplification of some patterns that require coordination between the RPC handler and a background task running on the emulated tile.  In a real device those RPCs may need to be implemented asynchronously since the rpc handler runs in interrupt context and cannot block but that can be tedious to reproduce in the emulator so this decorator allows a short-cut for implementing synchronous RPCs that need to temporarily yield or wait for whatever reason.

**Comment by timburke** [comment]
@DimaWittmann a reliable test case for this issue would need to be something like the following:

Custom emulated device with 2 emulated peripheral tiles: 1 and 2
- Tile 1 has an RPC Y that when called triggers it to send an RPC X to tile 2 from a background task
- Tile 2 implements RPC X asynchronously and does not complete it until it gets a specific flag set (which can be set directly on the tile class instance in the test case)
- Tile 1 and 2 also implement two other RPCs that are synchronous and let the proxy poll them to determine if they will return busy.

The test case then would be:
- Proxy calls RPC Y to start a long running background RPC
- Proxy then calls another synchronous RPC on tile 1 which succeeds
- Proxy then calls another synchronous RPC on tile 2 which should fail with busy
- Test case then sets flag on tile 2 to tell it to finish the asnyc rpc
- Proxy then calls another synchronous RPC on tile 2 which should succeed.

**Comment by DimaWittmann** [comment]
Thank you, Tim, for very thorough explanation, it makes everything clearer. 
The thing I miss is how to send automated RPC from another device. I tried creating a HardwareManager instance, connect, get proxy and call the RPC in the code, namely inside of RPC Y,  but I suppose it's wrong. May you, please, clarify how can I do it better?

**Comment by timburke** [comment]
@DimaWittmann 
The only way to send an automated RPC not from `HardwareManager` would be to create multiple EmulatedTile objects and add them to the same device.  There is an example here that has a single emulated peripheral tile:
https://github.com/iotile/coretools/blob/master/iotileemulate/iotile/emulate/demo/demo_device.py#L70

For the test case described in https://github.com/iotile/coretools/issues/619#issuecomment-465232755 you would need two emulated tiles in the same device.  

You can add as many tiles as you want to an EmulatedDevice in its constructor:
https://github.com/iotile/coretools/blob/master/iotileemulate/iotile/emulate/demo/demo_device.py#L192

To clarify, the primary way to send an rpc from **outside** of the emulated device is via HardwareManager and a proxy.  If you need to send another rpc for testing purposes, you need to do it in one of two ways:

- Have a tile **inside the device** send the rpc on your behalf as part of a background task running on the tile.  See `_application_main` [here](https://github.com/iotile/coretools/blob/master/iotileemulate/iotile/emulate/demo/demo_device.py#L88
).
- Have two threads and directly call the `rpc` method defined on `EmulatedDevice` from a background thread.  The public `rpc` method defined on EmulatedDevice gives you access to directly inject RPCs into the device's internal rpc queue from outside of the device.  Obviously the is a feature of the emulated device that physical devices don't possess.  The `rpc` method is declared [here](https://github.com/iotile/coretools/blob/master/iotileemulate/iotile/emulate/virtual/emulated_device.py#L129).

    This second method may be easier than the first but since `rpc` is a blocking call you would need a second thread in your test case in order to have two rpcs running at the same time without writing a custom `EmulatedPeripheralTile`.

**Comment by DimaWittmann** [comment]
Exactly, I see what you mean. 
I just wasn't sure how to call rpc of one tile in the other one but I suppose _self._device.emulator.await_rpc()_ should do the trick.

**Comment by timburke** [comment]
> Exactly, I see what you mean.
> I just wasn't sure how to call rpc of one tile in the other one but I suppose _self._device.emulator.await_rpc()_ should do the trick.

Yes correct, the way to send an rpc from inside of a tile is `await self._device.emulator.await_rpc()`

**Comment by DimaWittmann** [comment]
@timburke I was thinking how to handle a case when two serial RPC Y issued so another tile will receive two async rpcs. Should second call to RPC Y fail?

**Comment by timburke** [comment]
@DimaWittmann When any RPC is received (async or not, that answer isn't known yet at this point because the RPC handler has not yet been invoked to know if it would raise `AsynchronousRPCResponse`) the destination tile is checked to see if it currently has a pending async RPC.  If yes, then RPC is failed with busy, otherwise RPC handler on that tile is invoked.

Does that answer the question?

**Comment by DimaWittmann** [comment]
Tim, thank you very much for help. I created a PR and it would make me happy if you had a minute to take a look.
